src/analyses/mCP.ml

```diff
@@ -741,7 +731,17 @@ struct
     let sides  = ref [] in
     let assigns = ref [] in
     let emits = ref [] in
-    let f post_all (n,(module S:MCPSpec),d) =
+    let rec spec_list3_rev_acc acc l1 l2_opt l3 = match l1, l2_opt, l3 with
+      | [], _, [] -> acc
+      | (n, x) :: l1, Some ((n', y) :: l2), (n'', z) :: l3 when n = n' -> (* context-sensitive *)
+        assert (n = n'');
+        spec_list3_rev_acc ((n, spec n, (x, Some y, z)) :: acc) l1 (Some l2) l3
+      | (n, x) :: l1, l2_opt, (n'', z) :: l3 -> (* context-insensitive *)
+        assert (n = n'');
+        spec_list3_rev_acc ((n, spec n, (x, None, z)) :: acc) l1 l2_opt l3
+      | _, _, _ -> invalid_arg "MCP.spec_list3_rev_acc"
+    in
+    let f post_all (n,(module S:MCPSpec),(d,fc,fd)) =
```

Comment:
This is (at least to me) very hard to understand. Could you elaborate on what exactly happens here?

Reply:
It's like `spec_list2`, but for three lists, so the recursion here is basically what `map3_rev` would be. But here things are too nasty to have that because the context is optional (due to context insensitivity), so:
1. `MCP.combine` gets `fc: (int * Obj.t) list option`, so this handles the case when we don't have the second list at all. Then `None` is given to all inner analyses.
2. If that list is present, then there's still a second problem. If the analysis with ID `n` is in `ctx_insens`, then it has local domain elements from both the caller and the callee, but the corresponding element in the context list might have a different ID because it corresponds to already some following analysis. In that case also `None` is given to the inner analysis.
    Basically, the context list can be shorter than the local domain list and this matches them up by ID instead of doing something wrong.

Reply:
Thanks for explaining!